site/content/extend/plugins/signing-plugins.md

```diff
@@ -0,0 +1,159 @@
+```
+
+
+### Exporting the Private Key
+Find the ID of your private key first. The ID is a hexadecimal number.
+```console
+> gpg --list-secret-keys
+```
+This is your private key and you should keep it secret. Replace `XXXXXXXX` with your hexadecimal key ID
+```console
+> gpg --export-secret-keys XXXXXXXX > ./my-priv-key.plugin.gpg
+```
+
+
+### Exporting the Public Key
+Find the ID of your public key first. The ID is a hexadecimal number.
+```console
+> gpg --list-keys
+```
+Replace `XXXXXXXX` with your hexadecimal key ID
+```console
+> gpg --export XXXXXXXX > ./my-pub-key.plugin.gpg
+```
+
+
+### Importing the Key
+If you already have a public and private key pair, you can import them to the GPG.
+```console
+> gpg --import ./my-priv-gpg-key.plugin.gpg
+> gpg --import ./my-pub-gpg-key.plugin.gpg
+```
+
+
+## Plugin Signing
+For plugin signing you have to know the hexadecimal ID of the private key. Let's assume you want to sign `com.mattermost.demo-plugin-0.1.0.tar.gz` file. Replace `XXXXXXXX` with your hexadecimal key ID and run:
+```console
+> gpg -u XXXXXXXX --verbose --personal-digest-preferences SHA256 --detach-sign com.mattermost.demo-plugin-0.1.0.tar.gz
+```
+This command will generate `com.mattermost.demo-plugin-0.1.0.tar.gz.sig` which is the signature of your plugin.
+
+## Plugin Verification
+Mattermost server should be able to verify plugins before install. the public key has to be added on the server. For this purpose, start the Mattermost server and run the following command:
+```console
+> mattermost plugin add key my-pub-key.plugin.gpg
+```
+Note: if your public key file does not have an extension `.plugin.gpg` Mattermost will append it automatically.
+
+Multiple public keys can be added to the Mattermost server:
+```console
+> mattermost plugin add key my-pk-file1.plugin.gpg my-pk-file2.plugin.gpg
+```
+To list the names of all public keys installed on your Mattermost server, use:
+```console
+> mattermost plugin keys
+```
+To delete public key(s) from your Mattermost server, use:
+```console
+> mattermost plugin delete key my-pk-file1.plugin.gpg my-pk-file2.plugin.gpg
+```
+
+### Storing plugin signatures
+When a plugin is installed, Mattermost server will store plugin signatures in its internal filestore(not to be confused with the public keys). This is to allow the server to verify signatures during the plugin-synchronization process that happens during server startup. To avoid name collisions and overwriting existing files, Mattermost will store plugin signatures along side plugin bundles using the naming convention mentioned below. The benefit of storing signature files independently is to avoid double-tar, where the tar file contains the bundle tar along with its signature. This would have required Mattermost to migrate existing plugin bundles into double-tar format. Which would have resulted in increased complexity and made upgrade/downgrade scenarios a lot harder to deal with.
```

Comment:
```suggestion
When a plugin is installed, Mattermost server will store plugin signatures in its internal filestore (not to be confused with the public keys). This is to allow the server to verify signatures during the plugin-synchronization process that happens during server startup. To avoid name collisions and overwriting existing files, Mattermost will store plugin signatures alongside plugin bundles using the naming convention mentioned below. The benefit of storing signature files independently is to avoid double-tar, where the tar file contains the bundle tar along with its signature. This would have required Mattermost to migrate existing plugin bundles into double-tar format, which would have resulted in increased complexity and made upgrade/downgrade scenarios a lot harder to deal with.
```